### PR TITLE
NEXUS-7994 CMA ?describe

### PR DESCRIPTION
**Pull request by mrprescott**

### CODE REVIEW

**Comment by mrprescott** [comment]
https://issues.sonatype.org/browse/NEXUS-7994

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ConfigurableViewFacet.java [hunk 1]
```diff
@@ -52,11 +70,112 @@ protected void doConfigure() throws Exception {
...
+    }
   }
 
   @Override
   public boolean isOnline() {
     return online;
   }
+
+  private void describeRequest(final Description desc, final Request request) {
+    desc.topic("Request");
```
Comment (kellyrob99) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/ConfigurableViewFacet.java]: shame to have the fluent impl and not use it ;)

File: components/nexus-repository/src/main/resources/org/sonatype/nexus/repository/internal/describe/describe.vm [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+<head>
+  <title>DESCRIBE $path</title>
+  <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
+
+  <link rel="icon" type="image/png" href="$nexusUrl/favicon.png">
+  <!--[if IE]>
+  <link rel="SHORTCUT ICON" href="$nexusUrl/favicon.ico"/>
+  <![endif]-->
+
+  <!-- TODO Stylesheet -->
```
Comment (kellyrob99) [inline components/nexus-repository/src/main/resources/org/sonatype/nexus/repository/internal/describe/describe.vm]: is one actually required?

**Comment by kellyrob99** [comment]
+1 but could probably use some tests

**Comment by mrprescott** [comment]
http://bamboo.s/browse/NX3-OSSF81-3/ :white_check_mark:

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/internal/describe/TemplateDescriptionRenderer.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * Renders an HTML description of a {@link Description} with a Velocity template.
+ *
+ * @since 3.0
+ */
+@Named
+@Singleton
+public class TemplateDescriptionRenderer
+    implements DescriptionRenderer
+{
+  private static final String TEMPLATE_PATH = "describe.vm";
```
Comment (jdillon) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/internal/describe/TemplateDescriptionRenderer.java]: technically this is a path, its a resource name

**Comment by mrprescott** [comment]
Refactored it so that it's now httpbridge doing the ?describe handling.

File: plugins/basic/nexus-repository-httpbridge/src/main/java/org/sonatype/nexus/repository/httpbridge/internal/ViewServlet.java [hunk 4]
```diff
@@ -133,14 +148,52 @@ protected void doService(final HttpServletRequest httpRequest, final HttpServlet
...
+    else {
+      log.debug("HTTP response sender: {}", sender);
+      sender.send(response, httpResponse);
+    }
+  }
+
+  @VisibleForTesting
+  Response describe(final Request request, final Response response, final Exception exception) {
+    final Description description = new Description(ImmutableMap.<String, Object>of(
+        "path", nullToEmpty(request.getPath()),
```
Comment (jdillon) [inline plugins/basic/nexus-repository-httpbridge/src/main/java/org/sonatype/nexus/repository/httpbridge/internal/ViewServlet.java]: is request.getPath() ever null?

File: plugins/basic/nexus-repository-httpbridge/src/main/java/org/sonatype/nexus/repository/httpbridge/internal/describe/Description.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    }
+    return addTable(name, table);
+  }
+
+  public Map<String, Object> getParameters() {
+    return parameters;
+  }
+
+  public List<DescriptionItem> getItems() {
+    return Collections.unmodifiableList(items);
```
Comment (jdillon) [inline plugins/basic/nexus-repository-httpbridge/src/main/java/org/sonatype/nexus/repository/httpbridge/internal/describe/Description.java]: Any reason why we guard here, describe shouldn't modify these, just pass the items asis?

File: plugins/basic/nexus-repository-httpbridge/src/main/java/org/sonatype/nexus/repository/httpbridge/internal/describe/Description.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+/**
+ * Accumulates a renderable description of request-processing activity.
+ *
+ * @since 3.0
+ */
+public class Description
+{
+  private Map<String, Object> parameters;
+
+  private List<DescriptionItem> items = Lists.newArrayList();
```
Comment (jdillon) [inline plugins/basic/nexus-repository-httpbridge/src/main/java/org/sonatype/nexus/repository/httpbridge/internal/describe/Description.java]: appears to be final field

File: plugins/basic/nexus-repository-httpbridge/src/main/java/org/sonatype/nexus/repository/httpbridge/internal/describe/NullDescription.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * A do-nothing implementation of {@link Description}.
+ *
+ * @since 3.0
+ */
+public class NullDescription
+    extends Description
```
Comment (jdillon) [inline plugins/basic/nexus-repository-httpbridge/src/main/java/org/sonatype/nexus/repository/httpbridge/internal/describe/NullDescription.java]: This is not relevant anymore?

File: plugins/basic/nexus-repository-httpbridge/src/main/java/org/sonatype/nexus/repository/httpbridge/internal/describe/Description.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+  final private List<DescriptionItem> items = Lists.newArrayList();
+
+  public Description(final Map<String, Object> parameters) {
+    this.parameters = parameters;
+  }
+
+  public boolean isDescribing() {
+    return true;
+  }
```
Comment (jdillon) [inline plugins/basic/nexus-repository-httpbridge/src/main/java/org/sonatype/nexus/repository/httpbridge/internal/describe/Description.java]: Not relevant?

File: plugins/basic/nexus-repository-httpbridge/src/main/java/org/sonatype/nexus/repository/httpbridge/internal/describe/Description.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+  public Description topic(final String name) {
+    items.add(new DescriptionItem(name, "topic", name));
+    return this;
+  }
+
+  public Description add(final String name, final Object value) {
+    items.add(new DescriptionItem(name, "string", "" + value));
```
Comment (jdillon) [inline plugins/basic/nexus-repository-httpbridge/src/main/java/org/sonatype/nexus/repository/httpbridge/internal/describe/Description.java]: String.valueOf(value) is better than "" + value

**Comment by jdillon** [comment]
Using a model does add a bit of overhead to make collections for things.  It may be simpler to add the request and response to the template context and render, and leave the formatting of the content to the template vs. pre-rendering a model.

Not terribly concerned with the UT-ability of this, vs the simplicity of implementation.

FTR I think what we have now will work, but I am on the fence as to if its simple over potential alternatives.

**Comment by mrprescott** [comment]
http://bamboo.s/browse/NX3-OSSF81-4 :white_check_mark:

**Comment by jdillon** [comment]
+1

**Comment by mrprescott** [comment]
Manually squash-merged.

**Pull request closed without merging**